File: rental-manager-api/app/api/v1/endpoints/inventory/stock_levels.py
"""
API endpoints for stock level management.

Provides endpoints for viewing and managing inventory stock levels.
"""

from typing import Optional, List
from uuid import UUID
from decimal import Decimal
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db, get_current_user
from app.services.inventory.inventory_service import InventoryService
from app.schemas.inventory.stock_level import (
    StockLevelCreate,
    StockLevelUpdate,
    StockLevelResponse,
    StockLevelFilter,
    StockAdjustment,
    TransferRequest,
    StockSummaryResponse,
    LowStockAlert
)
from app.schemas.inventory.common import PaginatedResponse
from app.models.user import User


router = APIRouter()


@router.get("/", response_model=PaginatedResponse[StockLevelResponse])
async def list_stock_levels(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    item_id: Optional[UUID] = None,
    location_id: Optional[UUID] = None,
    low_stock_only: bool = False,
    include_zero: bool = True,
    db: AsyncSession = Depends(get_db)
):
    """
    List stock levels with optional filtering.
    
    Args:
        skip: Number of records to skip
        limit: Maximum number of records to return
        item_id: Filter by specific item
        location_id: Filter by specific location
        low_stock_only: Only show items below reorder point
        include_zero: Include items with zero stock
        
    Returns:
        Paginated list of stock levels
    """
    service = InventoryService()
    
    # Build filter
    filter_obj = StockLevelFilter(
        item_id=item_id,
        location_id=location_id,
        low_stock_only=low_stock_only,
        include_zero=include_zero
    )
    
    # Get filtered results
    from app.crud.inventory import stock_level as crud_stock_level
    
    query = await crud_stock_level.build_filter_query(
        db,
        filter_obj=filter_obj
    )
    
    total = await crud_stock_level.count(db, query=query)
    items = await crud_stock_level.get_multi(
        db,
        skip=skip,
        limit=limit,
        query=query
    )
    
    return PaginatedResponse(
        items=items,
        total=total,
        skip=skip,
        limit=limit
    )


@router.get("/summary", response_model=StockSummaryResponse)
async def get_stock_summary(
    item_id: Optional[UUID] = None,
    location_id: Optional[UUID] = None,
    category_id: Optional[UUID] = None,
    brand_id: Optional[UUID] = None,
    db: AsyncSession = Depends(get_db)
):
    """
    Get aggregated stock summary.
    
    Returns summary statistics for inventory including:
    - Total value
    - Total units
    - Stock by status
    - Low stock items count
    """
    service = InventoryService()
    
    summary = await service.get_stock_summary(
        db,
        item_id=item_id,
        location_id=location_id,
        category_id=category_id,
        brand_id=brand_id
    )
    
    return summary


@router.get("/alerts", response_model=List[LowStockAlert])
async def get_stock_alerts(
    location_id: Optional[UUID] = None,
    severity: Optional[str] = Query(None, regex="^(critical|warning|info)$"),
    db: AsyncSession = Depends(get_db)
):
    """
    Get inventory alerts.
    
    Returns alerts for:
    - Low stock items
    - Items needing maintenance
    - Expired warranties
    """
    service = InventoryService()
    
    alerts = await service.get_inventory_alerts(
        db,
        location_id=location_id,
        severity=severity
    )
    
    return alerts


@router.get("/{item_id}/{location_id}", response_model=StockLevelResponse)
async def get_stock_level(
    item_id: UUID,
    location_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    """
    Get stock level for specific item at location.
    
    Args:
        item_id: Item ID
        location_id: Location ID
        
    Returns:
        Stock level details
        
    Raises:
        404: Stock level not found
    """
    from app.crud.inventory import stock_level as crud_stock_level
    
    stock_level = await crud_stock_level.get_by_item_location(
        db,
        item_id=item_id,
        location_id=location_id
    )
    
    if not stock_level:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Stock level not found for item {item_id} at location {location_id}"
        )
    
    return stock_level


@router.post("/initialize", response_model=StockLevelResponse)
async def initialize_stock_level(
    stock_data: StockLevelCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    Initialize stock level for an item at a location.
    
    Args:
        stock_data: Initial stock information
        
    Returns:
        Created stock level
        
    Raises:
        409: Stock level already exists
    """
    service = InventoryService()
    
    try:
        stock_level = await service.initialize_stock_level(
            db,
            item_id=stock_data.item_id,
            location_id=stock_data.location_id,
            initial_quantity=stock_data.quantity_on_hand,
            reorder_point=stock_data.reorder_point,
            reorder_quantity=stock_data.reorder_quantity,
            created_by=current_user.id
        )
        
        await db.commit()
        return stock_level
        
    except ValueError as e:
        if "already exists" in str(e):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=str(e)
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.put("/{stock_level_id}", response_model=StockLevelResponse)
async def update_stock_level(
    stock_level_id: UUID,
    update_data: StockLevelUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    Update stock level configuration.
    
    Updates reorder points and other configuration.
    Does NOT update quantities - use adjustment endpoint for that.
    
    Args:
        stock_level_id: Stock level ID
        update_data: Update data
        
    Returns:
        Updated stock level
    """
    from app.crud.inventory import stock_level as crud_stock_level
    
    stock_level = await crud_stock_level.get(db, id=stock_level_id)
    
    if not stock_level:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Stock level not found"
        )
    
    # Update fields
    update_dict = update_data.model_dump(exclude_unset=True)
    update_dict["updated_by"] = current_user.id
    
    stock_level = await crud_stock_level.update(
        db,
        db_obj=stock_level,
        obj_in=update_dict
    )
    
    await db.commit()
    return stock_level


@router.post("/adjust", response_model=StockLevelResponse)
async def adjust_stock(
    adjustment: StockAdjustment,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    Perform stock adjustment.
    
    Used for:
    - Inventory counts
    - Corrections
    - Write-offs
    - Found inventory
    
    Args:
        adjustment: Adjustment details
        
    Returns:
        Updated stock level
    """
    service = InventoryService()
    
    try:
        stock_level = await service.perform_stock_adjustment(
            db,
            item_id=adjustment.item_id,
            location_id=adjustment.location_id,
            adjustment_type=adjustment.adjustment_type,
            quantity=adjustment.quantity,
            reason=adjustment.reason,
            reference_number=adjustment.reference_number,
            performed_by=current_user.id
        )
        
        await db.commit()
        return stock_level
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/transfer", response_model=dict)
async def transfer_stock(
    transfer: TransferRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    Transfer stock between locations.
    
    Args:
        transfer: Transfer details
        
    Returns:
        Transfer confirmation with movement IDs
    """
    service = InventoryService()
    
    try:
        from_movement, to_movement = await service.transfer_stock(
            db,
            item_id=transfer.item_id,
            from_location_id=transfer.from_location_id,
            to_location_id=transfer.to_location_id,
            quantity=transfer.quantity,
            reason=transfer.reason,
            transferred_by=current_user.id
        )
        
        await db.commit()
        
        return {
            "status": "success",
            "from_movement_id": str(from_movement.id),
            "to_movement_id": str(to_movement.id),
            "quantity_transferred": float(transfer.quantity)
        }
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/rental/checkout", response_model=dict)
async def process_rental_checkout(
    item_id: UUID = Query(...),
    location_id: UUID = Query(...),
    quantity: Decimal = Query(..., gt=0),
    customer_id: UUID = Query(...),
    transaction_id: Optional[UUID] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    Process rental checkout.
    
    Allocates inventory units for rental.
    
    Args:
        item_id: Item to rent
        location_id: Rental location
        quantity: Quantity to rent
        customer_id: Customer ID
        transaction_id: Optional transaction reference
        
    Returns:
        Checkout confirmation with allocated units
    """
    service = InventoryService()
    
    try:
        units, movement = await service.process_rental_checkout(
            db,
            item_id=item_id,
            location_id=location_id,
            quantity=quantity,
            customer_id=customer_id,
            transaction_id=transaction_id,
            processed_by=current_user.id
        )
        
        await db.commit()
        
        return {
            "status": "success",
            "units_allocated": len(units),
            "unit_ids": [str(unit.id) for unit in units],
            "movement_id": str(movement.id)
        }
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/rental/return", response_model=dict)
async def process_rental_return(
    unit_ids: List[UUID] = Query(...),
    location_id: UUID = Query(...),
    condition: Optional[str] = Query("good"),
    damage_notes: Optional[str] = None,
    transaction_id: Optional[UUID] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    Process rental return.
    
    Returns rented units to inventory.
    
    Args:
        unit_ids: List of unit IDs to return
        location_id: Return location
        condition: Condition of returned items
        damage_notes: Notes about any damage
        transaction_id: Optional transaction reference
        
    Returns:
        Return confirmation
    """
    service = InventoryService()
    
    try:
        movement = await service.process_rental_return(
            db,
            unit_ids=unit_ids,
            location_id=location_id,
            condition=condition,
            damage_notes=damage_notes,
            transaction_id=transaction_id,
            processed_by=current_user.id
        )
        
        await db.commit()
        
        return {
            "status": "success",
            "units_returned": len(unit_ids),
            "movement_id": str(movement.id),
            "condition": condition
        }
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get("/availability/check", response_model=dict)
async def check_availability(
    item_id: UUID = Query(...),
    location_id: UUID = Query(...),
    quantity: Decimal = Query(..., gt=0),
    start_date: Optional[datetime] = None,
    end_date: Optional[datetime] = None,
    db: AsyncSession = Depends(get_db)
):
    """
    Check item availability.
    
    Args:
        item_id: Item to check
        location_id: Location to check
        quantity: Required quantity
        start_date: Rental start date (for future reservations)
        end_date: Rental end date
        
    Returns:
        Availability information
    """
    from app.crud.inventory import stock_level as crud_stock_level
    
    stock_level = await crud_stock_level.get_by_item_location(
        db,
        item_id=item_id,
        location_id=location_id
    )
    
    if not stock_level:
        return {
            "available": False,
            "quantity_available": 0,
            "message": "Item not stocked at this location"
        }
    
    available = stock_level.quantity_available >= quantity
    
    return {
        "available": available,
        "quantity_available": float(stock_level.quantity_available),
        "quantity_requested": float(quantity),
        "quantity_on_rent": float(stock_level.quantity_on_rent),
        "quantity_reserved": float(stock_level.quantity_reserved)
    }

File: rental-manager-frontend/src/services/api/stock-levels.ts
/**
 * API service for Stock Levels management
 */
import { apiClient } from '@/lib/axios';

export interface StockLevel {
  id: string;
  item_id: string;
  item_name: string;
  item_sku: string;
  location_id: string;
  location_name: string;
  quantity_on_hand: number;
  quantity_available: number;
  quantity_reserved: number;
  quantity_on_rent: number;
  quantity_in_maintenance: number;
  quantity_damaged: number;
  reorder_point: number | null;
  maximum_stock: number | null;
  total_value: number;
  last_updated: string;
  stock_status: 'IN_STOCK' | 'LOW_STOCK' | 'OUT_OF_STOCK';
  category?: {
    id: string;
    name: string;
    code: string;
  };
  brand?: {
    id: string;
    name: string;
  };
}

export interface StockSummary {
  total_on_hand: number;
  total_available: number;
  total_reserved: number;
  total_on_rent: number;
  total_damaged: number;
  total_value: number;
  location_count: number;
  item_count: number;
  low_stock_count: number;
  utilization_rate: number;
  availability_rate: number;
}

export interface LowStockAlert {
  id: string;
  item_id: string;
  item_name: string;
  item_sku: string;
  location_id: string;
  location_name: string;
  current_stock: number;
  reorder_point: number;
  shortage: number;
  severity: 'high' | 'medium' | 'low';
  alert_type: 'LOW_STOCK' | 'OUT_OF_STOCK';
}

export interface StockAdjustmentRequest {
  item_id: string;
  location_id: string;
  adjustment_type: 'ADD' | 'REMOVE' | 'SET';
  quantity: number;
  reason: string;
  notes?: string;
  cost_per_unit?: number;
}

export interface StockTransferRequest {
  item_id: string;
  from_location_id: string;
  to_location_id: string;
  quantity: number;
  transfer_notes?: string;
  requires_approval?: boolean;
}

export interface GetStockLevelsParams {
  search?: string;
  location_id?: string;
  item_id?: string;
  category_id?: string;
  brand_id?: string;
  stock_status?: 'IN_STOCK' | 'LOW_STOCK' | 'OUT_OF_STOCK';
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  skip?: number;
  limit?: number;
}

export const stockLevelsApi = {
  /**
   * Get all stock levels with filtering
   */
  async getStockLevels(params?: GetStockLevelsParams): Promise<{data: StockLevel[], total: number}> {
    const queryParams = new URLSearchParams();
    
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null && value !== '') {
          queryParams.append(key, String(value));
        }
      });
    }
    
    const response = await apiClient.get<{success: boolean; data: StockLevel[]; total: number}>(
      `/inventory/stock-levels?${queryParams.toString()}`
    );
    
    if (response.data.success) {
      return {
        data: response.data.data || [],
        total: response.data.total || 0
      };
    }
    
    return { data: [], total: 0 };
  },

  /**
   * Get stock summary statistics
   */
  async getStockSummary(location_id?: string): Promise<StockSummary> {
    const params = location_id ? `?location_id=${location_id}` : '';
    const response = await apiClient.get<{success: boolean; data: StockSummary}>(
      `/inventory/stock-levels/summary${params}`
    );
    
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    
    // Return default summary if no data
    return {
      total_on_hand: 0,
      total_available: 0,
      total_reserved: 0,
      total_on_rent: 0,
      total_damaged: 0,
      total_value: 0,
      location_count: 0,
      item_count: 0,
      low_stock_count: 0,
      utilization_rate: 0,
      availability_rate: 0,
    };
  },

  /**
   * Get low stock alerts
   */
  async getLowStockAlerts(): Promise<LowStockAlert[]> {
    const response = await apiClient.get<{success: boolean; data: LowStockAlert[]}>(
      '/inventory/stock-levels/alerts'
    );
    
    if (response.data.success) {
      return response.data.data || [];
    }
    
    return [];
  },

  /**
   * Get specific stock level for item and location
   */
  async getStockLevel(item_id: string, location_id: string): Promise<StockLevel | null> {
    try {
      const response = await apiClient.get<{success: boolean; data: StockLevel}>(
        `/inventory/stock-levels/${item_id}/${location_id}`
      );
      
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
    } catch (error) {
      console.error(`Error fetching stock level for item ${item_id} at location ${location_id}:`, error);
    }
    
    return null;
  },

  /**
   * Initialize stock level for item at location
   */
  async initializeStockLevel(data: {
    item_id: string;
    location_id: string;
    initial_quantity?: number;
    reorder_point?: number;
    maximum_stock?: number;
  }): Promise<StockLevel> {
    const response = await apiClient.post<{success: boolean; data: StockLevel}>(
      '/inventory/stock-levels/initialize',
      data
    );
    
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    
    throw new Error('Failed to initialize stock level');
  },

  /**
   * Perform stock adjustment
   */
  async adjustStock(adjustment: StockAdjustmentRequest): Promise<StockLevel> {
    const response = await apiClient.post<{success: boolean; data: StockLevel}>(
      '/inventory/stock-levels/adjust',
      adjustment
    );
    
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    
    throw new Error('Failed to adjust stock');
  },

  /**
   * Transfer stock between locations
   */
  async transferStock(transfer: StockTransferRequest): Promise<{success: boolean; message: string}> {
    const response = await apiClient.post<{success: boolean; message: string; data?: any}>(
      '/inventory/stock-levels/transfer',
      transfer
    );
    
    return {
      success: response.data.success,
      message: response.data.message || 'Transfer completed'
    };
  },

  /**
   * Update stock level settings (reorder point, max stock)
   */
  async updateStockLevel(stock_level_id: string, data: {
    reorder_point?: number;
    maximum_stock?: number;
  }): Promise<StockLevel> {
    const response = await apiClient.put<{success: boolean; data: StockLevel}>(
      `/inventory/stock-levels/${stock_level_id}`,
      data
    );
    
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    
    throw new Error('Failed to update stock level');
  },

  /**
   * Check availability for rental/sale
   */
  async checkAvailability(data: {
    item_id: string;
    location_id?: string;
    quantity: number;
    start_date?: string;
    end_date?: string;
  }): Promise<{available: boolean; available_quantity: number; message?: string}> {
    const response = await apiClient.get<{success: boolean; data: any}>(
      `/inventory/stock-levels/availability/check?${new URLSearchParams(
        Object.entries(data).filter(([_, value]) => value !== undefined).map(([key, value]) => [key, String(value)])
      ).toString()}`
    );
    
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    
    return {
      available: false,
      available_quantity: 0,
      message: 'Unable to check availability'
    };
  }
};